Document pid_file_name and explain some other flags (#243)
spiffe/spiffe-helper: README.md

If `-config` is not specified, the default value `helper.conf` is assumed. 

CLI options:

 | Flag name                       | Description                                                         |
 |---------------------------------|---------------------------------------------------------------------|
 | `-config`                       | Path to the configuration file                                      |
 | `-help`                         | Print interactive help                                              |

>  | `-daemon-mode={true|false}`     | Overrides `daemon_mode` in the config file.                         |

Reviewer: I think you need to escape the pipe. This line is not rendering properly

Author: Good catch. Fixed

If `-config` is not specified, the default value `helper.conf` is assumed. 

CLI options:

 | Flag name                       | Description                                                         |
 |---------------------------------|---------------------------------------------------------------------|
 | `-config`                       | Path to the configuration file                                      |
 | `-help`                         | Print interactive help                                              |
 | `-daemon-mode={true|false}`     | Overrides `daemon_mode` in the config file.                         |

>  | `-exitWhenReady`                | Obsolete and ignored. Use `-daemon-mode=false` instead.             |

Reviewer 2: We should probably just delete this rather than documenting it.

Author: PR to delete the Config field: https://github.com/spiffe/spiffe-helper/pull/248

 |-------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
 | `agent_address`                     | Socket address of SPIRE Agent.                                                                                                    | `"/tmp/agent.sock"`                                                                                                                                                  |
 | `cmd`                               | The path to the process to launch.                                                                                                | `"ghostunnel"`                                                                                                                                                       |

>  | `cmd_args`                          | The arguments of the process to launch and (in daemon mode) monitor and signal for certificate renewals.                          | `"server --listen localhost:8002 --target localhost:8001--keystore certs/svid_key.pem --cacert certs/svid_bundle.pem --allow-uri-san spiffe://example.org/Database"` |

Reviewer: I think the part about monitoring and signaling would work better as a description for cmd

Author: Updated

when the certificates are reloaded.

#### Use in daemon-mode with file watcher

If the process using the certificates can watch the on-disk files for changes,
spiffe-helper can be run in `daemon_mode` with no `cmd` or `pid_file_name`. It
will fetch and renew the certificates, overwrite them on disk, and the process
will pick up the changes.

This approach is recommended where possible.

#### Use in daemon-mode with `cmd` to run a process or a reload command

`cmd` and `cmd_args` are used in `daemon_mode` to run a command whenever the
certificates are renewed. This can be a long-lived process that uses the
certificates, or a short-lived command that signals a reload mechanism
for an externally-managed process.

:warning: **cmd_args is not parsed according to shell-like rules**. The
`cmd_args` will be split into individual arguments using space separation
unless the argument is enclosed in double quotes, which are consumed. Double
quotes must be backslash escaped in the hcl string. For example:

```hcl
cmd_args = "\"this is one argument\""
```

Double quotes within the argument string must be escaped by doubling them so
they are not interpreted as argument delimiters. E.g.:

```hcl

> cmd_args = "\"this is a single argument with ONE double-quote \"\" in it\""

Reviewer: Thanks for updated the explanation.  To have a single double quote it wouldn't be `\\\"`?

Author: No - see the test cases. The CSV parser consumes `""` -> `"`, not `\"` -> `"`

Reviewer: Thanks for the explanation

 | `jwt_bundle_file_mode`              | The octal file mode to use when saving a JWT Bundle file.                                                                         | `0600`                                                                                                                                                               |
 | `jwt_svid_file_mode`                | The octal file mode to use when saving a JWT SVID file.                                                                           | `0600`                                                                                                                                                               |
### Operating modes and configuration details

spiffe-helper has two primary operating modes - "daemon mode" (the default),
where it runs continuously and manages the certificates, and "non-daemon mode",
where it fetches the certificates once and exits.

The helper can be used in several ways, as detailed below:

> * To run a command that uses the certificates, and signal it when they are renewed.
> * To run a command that reloads an external process when the certificates are renewed.
> * To signal an external process when the certificates are renewed.
> * To fetch the certificates once and exit.

Reviewer: ```suggestion
* Run a command that uses the certificates, and signal it when they are renewed.
* Run a command that reloads an external process when the certificates are renewed.
* Signal an external process when the certificates are renewed.
* Fetch the certificates once and exit.
```

process and background itself like a unix "daemon". It just keeps running until
terminated by a signal or a fatal error.

#### Use in daemon-mode with file watcher

If the process using the certificates can watch the on-disk files for changes,
spiffe-helper can be run in `daemon_mode` with no `cmd` or `pid_file_name`. It
will fetch and renew the certificates, overwrite them on disk, and the process
will pick up the changes.

This approach is recommended where possible.

#### Use in daemon-mode with `cmd` to run a process or a reload command

`cmd` and `cmd_args` are used in `daemon_mode` to run a command whenever the
certificates are renewed. This can be a long-lived process that uses the
certificates, or a short-lived command that signals a reload mechanism
for an externally-managed process.

The `cmd_args` will be split into individual arguments using space separation
unless the argument is enclosed in double quotes, which are consumed. Single quotes
are NOT respected for argument quoting, so `cmd` = `touch` and `cmd_args` =
`'some file'` will create two files named `'some` and `file'` (including the quotes

> in the file name), not a single file named `some file`.

Author: The problem is that `spiffe-helper` uses a csv parser with the delimiter set to " " (space) to process its command-line args and split them to an argument vector.

This is the test case I wrote that this was based on: https://github.com/ringerc/spiffe-helper-PRs/blob/31a134b2438ea865739673779c23c99beb410815/pkg/sidecar/sidecar_test.go#L116-L131

This would be correct as a shell argument string, but is not handled correctly by spiffe-helper because spiffe-helper only supports double-quoted arguments.

To confirm my understanding I've banged out some test cases here https://github.com/spiffe/spiffe-helper/pull/256 - see in particular https://github.com/spiffe/spiffe-helper/pull/256#pullrequestreview-2594036622 for this case.

Using a [csv parser](https://github.com/spiffe/spiffe-helper/blob/8d77e4a5a16d623448cf8228b21e97c8defd8e26/pkg/sidecar/sidecar.go#L412-L425) produces surprising and confusing results, as the semantics of csv are not the same as the semantics of a command line argument string, particularly when nesting double quotes within single quotes (which in the shell's parser, protects the double quotes).

This could result in surprising and even dangerous behaviour.

I [implemented an argument-vector replacement](https://github.com/ringerc/spiffe-helper-PRs/blob/dce4481e0d7b2d35dfaf14556afee3e84add7b07/cmd/spiffe-helper/config/config.go#L99-L126) in my working tree but I haven't had time to extract it into a separate PR, and I'd rather wait until some of the current ones are merged so it doesn't get too hard to keep track of things and rebase.

So I documented the current behaviour in the meantime. Would it be clearer if I split this into whole hcl lines?

> in the file name), not a single file named `some file`.

Author: Revised the explanation to hopefully make it clearer at the expense of making it longer: https://github.com/spiffe/spiffe-helper/pull/243/commits/a5582541d279a12ff9633682f96e8bc789a15671

invocations, as they can introduce security vulnerabilities and should be
avoided where possible.

The command's stdout and stderr will be attached to spiffe-helper's stdout
and stderr. Its stdin will be closed.

The process will not be launched for the first time until the certificates
are fetched successfully.

`spiffe-helper` continues running if the process created by `cmd` exits. The
process is not re-launched immediately if it exits.

If the process is still running next time the certs are renewed,
`spiffe-helper` will signal it with `renew_signal`. If it has exited,
`spiffe-helper` will re-launch it instead. This can be used to manage
certificate reloading in externally-managed processes that do not support
reloading certificates with a signal.

#### Use in daemon-mode with `pid_file_name` to signal an externally-managed process

If running in `daemon_mode` with `pid_file_name` set, the pid in
`pid_file_name` is sent the signal `renew_signal` to reload the certificates
when they are renewed.

`pid_file_name` is re-read every time the process is to be signaled, so it can
be updated with a new pid if the process changes.

> An error will be logged if the pid file cannot be opened, read or the value parsed,

Reviewer: ```suggestion
An error will be logged if the pid file cannot be opened, read, or the value parsed,
```

and the attempt to signal the process will be skipped. The process will still be
signalled next time the certificates are renewed.

#### Combining `cmd` and `pid_file_name`

Both `cmd` and `pid_file_name` can be used at the same time. `spiffe-helper` will
both run the specified command and signal the process in the pid file. This can be
useful to use `cmd` to start the process to be managed, and `pid_file_name` to signal
a reload mechanism that is not signal-based.

#### Use in one-shot non-daemon mode

If `daemon_mode` is false, `spiffe-helper` will fetch the certificates once and
exit. This can be useful for one-shot scripts or for use in a process supervisor
that does not require a long-lived process.

> `cmd` is ignored in non-daemon mode. No command will be run.

Reviewer: ```suggestion
`cmd` and `pid_file_name` are ignored in non-daemon mode. No command will be run.
```

Author: Applied

 | `jwt_bundle_file_mode`              | The octal file mode to use when saving a JWT Bundle file.                                                                         | `0600`                                                                                                                                                               |
 | `jwt_svid_file_mode`                | The octal file mode to use when saving a JWT SVID file.                                                                           | `0600`                                                                                                                                                               |

> 

Reviewer: Notes:

> 

Reviewer: ```suggestion
Notes:
```

> `cmd` and `cmd_args` are used in `daemon_mode` to launch the process that will

Reviewer: Can we convert these to an unordered list?

Author: Done
